### PR TITLE
Split media and config into separate partitions

### PR DESCRIPTION
**Pull request by jasaw**
Add a new config partition for etc, log, varlib. Config partition mounted as synchronous to minimize filesystem corruption in the event of power loss.
Media files go into media storage partition.

Newly burned images will start with separate config and media partitions. Devices upgrading from old firmware will maintain the combined config and media partition.

Also reduce logrotate max log file size from 10M to a more sensible 3M.

This is the first step to addressing issue #1164.

What's lacking:
- motionEye to prevent saving media files to config partition.

I know this is a ThingOS issue, but I just want to get the discussion started first (also lack of free time). I can move this to ThingOS later if it makes things easier.

### CODE REVIEW

**Comment by ccrisan** [comment]
TL;DR

I really think this is not the best way to go about the issue at hand here. I appreciate your work on this matter, however here are my concerns/ideas:
 1. Firmware upgrades will practically never allow this setup unless we do some serious voodoo at first post-upgrade boot. This would leave many users with an old setup.
 2. This is indeed something that should go to thingOS (as you have mentioned) but I personally don't think it can be generalized enough to fit thingOS and all its applications. Take for example a project that stores its configs (and other stuff) in a DB - it's not the case for motionEye but it is for some of my own projects based on thingOS. DBs can grow and it'll become hard to draw a line between what is "config" and what is "data".
 3. With regards to video surveillance, I wouldn't generally recommend saving these videos/pictures to an SD card that runs the OS, unless it's temporary (i.e. before being uploaded to cloud, ftp or so). The recommended way is to use an external hard drive. Reasons include available space, (sometimes) speed, reliability and lifetime.
 4. Users already have issues with dealing with 3 partitions, some of them not visible from Windows, some of them being writable, some not. Having yet another partition may cause even more confusion.
 5. The main application of the OS, whatever it is, should never allow running out of disk space. We have control on the disk space and we should properly implement this in motionEye.
 6. Other possible disk filling OS parts are log files which get rotated (as you pointed out) and can be [moved to RAM](https://github.com/ccrisan/thingos/wiki/Logs#logs-in-ram) if that is a concern.
 7. The probability of file corruption is basically the same. Having the config partition mounted as "sync" is a good idea indeed, but it can also be achieved in a way by explicitly calling the "sync" command (or syscall) after altering the config, even if it lives on the same data partition.

I'm not saying there wouldn't be advantages if going this way and having two writable partitions. All I'm saying here is that I don't think it's worth the trouble and I believe we're trying to work around the main problem (which is filling up the disk by media files) using a rather radical OS change.

**Comment by ccrisan** [comment]
> The log files are extremely valuable when it comes to debugging issues, so they need to be persistent. Storing it in RAM is definitely a no no.

That's why they go to disk by default. However a few users have expressed interest in having them in RAM, just to minimize writes to the card. So they have a choice now.

> Yes, we should explore this option too. Do you have an idea on what you might implement in motionEye?

Of course. My idea is to implement the following:
 * make the disk progress go red or something to signal approaching the disk limit (limit to be determined); a popup when visiting the UI could also be implemented
 * add an option to the Preserve setting(s) to allow automatic removal of older media files when the disk limit is approached
 * hit the user with a big fucking warning about that, since we're deleting user's files
 * instruct the motion daemon to automatically stop creating media files when the disk usage becomes critical (limit to be determined)

**Comment by jasaw** [comment]
> make the disk progress go red or something to signal approaching the disk limit (limit to be determined); a popup when visiting the UI could also be implemented

Nice to have, but I imagine most users will not visit the web UI unless something broke, or changing configuration.

> add an option to the Preserve setting(s) to allow automatic removal of older media files when the disk limit is approached

If we are keeping the media and config as one partition, I would make automatic removal of old media mandatory. I would rename "Preserve" option to "until disk full" to let users know that their files WILL be deleted automatically. Also warn the user about the fact (your point below).

> hit the user with a big fucking warning about that, since we're deleting user's files

Yes, big fat warning.

> instruct the motion daemon to automatically stop creating media files when the disk usage becomes critical (limit to be determined)

I wouldn't stop motion from creating new media. Since we are deleting old media, I like to think of the media storage as one big circular buffer. This way, the camera keeps working, and the media files stay on the partition for months before getting removed automatically. Also, like you said, users should have uploaded the media somewhere else anyway for various reasons.


Your solution sounds workable. I'll close this PR.

**Comment by ccrisan** [comment]
>> make the disk progress go red or something to signal approaching the disk limit (limit to be determined); a popup when visiting the UI could also be implemented

>Nice to have, but I imagine most users will not visit the web UI unless something broke, or changing configuration.

Nevertheless it's something that would draw your attention if you happen to visit the UI.

>> instruct the motion daemon to automatically stop creating media files when the disk usage becomes critical (limit to be determined)

> I wouldn't stop motion from creating new media. Since we are deleting old media, I like to think of the media storage as one big circular buffer. This way, the camera keeps working, and the media files stay on the partition for months before getting removed automatically. Also, like you said, users should have uploaded the media somewhere else anyway for various reasons.

This is meant as an alternative to the "preserve until disk full" feature, which is indeed supposed to work as a circular... buffer (or something). If the user fails to activate that feature (which will be deactivated by default, since we're not erasing user's movies without consent), we must do something when the disk usage reaches a critical limit (99% or so). And that thing is to stop creating even more media files, since it's not going to work anyway. I'm not 100% sure about this but I think it might be worth implementing.

Another idea that came to me these days is to change the `text_overlay` of the cameras into a WARNING when the disk limit becomes critical. This way, even if you don't visit the UI often, you'll get to see the warning eventually on the streams (that you supposedly watch every now and then).

**Comment by BCGuy** [comment]
Is it possible to change the permission on the logs or other administrative files so they could be copied and then deleted by the user thereby freeing up enough space to boot?  Or just put in a dummy file that could be deleted and be reinstated after a clean up procedure?  

The work down on this project is amazing.

Regards,
A. Noob

**Comment by jasaw** [comment]
@BCGuy As a workaround until the proper solution is implemented, I split the media and logs into their respective partitions, exactly as described in this pull-request. It has been working well for me so far.
